core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -498,6 +498,112 @@ class RemoteLogManager(fetchLog: TopicPartition => Option[Log],
         updateRemoteLogStartOffset(topicPartition, remoteLogStartOffset)
       }
 
+      /*
+      A segment has leader epochs (remote leader epochs). A partition leader has leader epochs (local leader epochs).
+      For a segment to have a good lineage its leader epochs must be within the boundaries of the leader epochs
+      of the partition leader.
+      The first leader epoch in the segment always starts at the first offset of
+      the segment even if the partition leader's leader epoch started at an earlier log offset.
+      The last leader epoch in the segment always finishes at the last offset of
+      the segment even if the partition leader's leader epoch finishes at a further log offset.
+
+      CORRECT
+      Remote:             i1 ===================== i2
+      Local :        <=== j1 ======================================== j2 ===>
+
+      CORRECT
+      Remote:                                i1 ===================== i2
+      Local :        <=== j1 ======================================== j2 ===>
+
+      CORRECT
+      Remote:                      i1 ===================== i2
+      Local :        <=== j1 ======================================== j2 ===>
+
+      The below is a situation where the remote segment has the same leader epoch as the currently
+      ongoing local leader epoch.
+      CORRECT
+      Remote:                                i1 ===================== i2
+      Local :        <=== j1 ===============================================>
+
+      INCORRECT
+      Remote:                                       i1 ===================== i2
+      Local :        <=== j1 ======================================== j2 ===>
+
+      INCORRECT
+      Remote:   i1 ===================== i2
+      Local :        <=== j1 ======================================== j2 ===>
+      */
+      def doesRemoteSegmentHaveGoodLineage(validEpochs: util.TreeMap[Integer, lang.Long],
+                                           segmentMetadata: RemoteLogSegmentMetadata,
+                                           logEndOffset: lang.Long): Boolean = {
+        val segmentEpochs = segmentMetadata.segmentLeaderEpochs()
+        val segmentEndOffset = segmentMetadata.endOffset()
+        val firstSegmentEpochEntry = segmentEpochs.firstEntry()
+        segmentEpochs.entrySet().stream()
+          .allMatch(entry => {
+            val segmentEpochKey = entry.getKey
+            val segmentEpochStartOffset = entry.getValue
+            if (validEpochs.containsKey(segmentEpochKey)) {
+              val validEpochStartOffset = validEpochs.get(segmentEpochKey)
+              val validEpochEndEntry = validEpochs.higherEntry(segmentEpochKey)
+              val remoteEndEntry = segmentEpochs.higherEntry(segmentEpochKey)
+
+              if (validEpochEndEntry == null && remoteEndEntry != null) {
+                // Remote has more entries than local
+                // This should be impossible but we are erring on the side of defining the behaviour
```

Comment:
Correct me if I am wrong here Satish but from what I understand leader epoch increases on every leader election i.e. it is guaranteed to increase monotonically. Another way to say this is that there can be no higher epoch in the cluster at any time other than the largest epoch in leader epoch history.

Even in case of unclean leader election, the new leader (an out of sync follower) will have an epoch which is larger than any other in the cluster at any time. Hence, we cannot have a segment with an epoch which is larger than the largest epoch on current leader epoch history/lineage.

Reply:
Right. I misread the value `validEpochEndEntry` computed. The current leader epoch is always more than the earlier leader epochs.